Output and record standard streams in in-toto-run

lukpueh opened the pull request:
**Fixes issue #**:
Fixes #11 

**Description of the changes being introduced by the pull request**:
Add `subprocess.run`-like function to `in_toto.process` module that redirects standard output and standard error of the executed command to temporary files and both writes the contents to the corresponding streams of the parent process during execution and also returns them along
with the child process's return code upon its termination.

This function is called `in_toto_run` when passing `record_streams=True`.

The PR is a friendly takeover of #160. Thanks for the solid ground work, @ng1355! 

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
Add subprocess run function that duplicates streams

Add run-like function to in_toto.process module that redirects
standard output and standard error of executed command to temporary
files and both writes the contents to the corresponding streams of
the parent process during execution and also returns them along
with the child process's return code upon its termination.

Commit:
Duplicate streams when passing record_streams

Use newly added `process.run_duplicate_streams` function to
both write a child process's standard streams to the corresponding
parent process streams and store them in the generated link
metadata, when calling in-toto-run with the record_streams flag.

SantiagoTorres commented on in_toto/process.py:
```diff
@@ -40,11 +45,11 @@
 PIPE = subprocess.PIPE
 
 
+
```
Probably this stray newline can go away.

SantiagoTorres commented on in_toto/process.py:
```diff
@@ -118,3 +123,97 @@ def run(cmd, check=True, timeout=SUBPROCESS_TIMEOUT, **kwargs):
+        sys.stdout.write(stdout_part)
+        sys.stderr.write(stderr_part)
+        sys.stdout.flush()
+        sys.stderr.flush()
```
silly comment, but are all of these IPC calls available on windows too?

lukpueh replied:
According to the passing windows tests on appveyor they do.

SantiagoTorres commented on tests/test_process.py:
```diff
@@ -50,5 +54,67 @@ def test_input_vs_stdin(self):
+  def test_run_duplicate_streams_timeout(self):
+    """Test raise TimeoutExpired. """
+    with self.assertRaises(in_toto.process.subprocess.TimeoutExpired):
+      in_toto.process.run_duplicate_streams("python --version", timeout=-1)
```
how is this timing out?

lukpueh replied:
It times out a second before it started, and thus returns immediately. `subprocess`' timeout has the same behavior.

Commit:
Close files used by subprocess in windows

Just passing the file descriptor as stdout/stderr to a subprocess
makes windows complain that the subprocess is still using the
file after returning. To fix this we can open the file in the
parent process (inside a with statement) and pass the file handle
to Popen.
